File: src/client/tk/quickentry/QuickEntrySubscriber.py
import logging
import threading
from typing import TYPE_CHECKING, Callable

if TYPE_CHECKING:
    from src.types import RecognitionResult


class QuickEntrySubscriber:

    def __init__(
        self,
        on_text_change: Callable[[str, str], None],
        verbose: bool = False
    ) -> None:
        self._on_text_change = on_text_change
        self._verbose = verbose
        self._lock = threading.Lock()
        self._finalized_text = ""
        self._partial_text = ""
        self._active = False

    def activate(self) -> None:
        with self._lock:
            self._active = True
            self._finalized_text = ""
            self._partial_text = ""

        if self._verbose:
            logging.info("QuickEntrySubscriber: activated")

    def deactivate(self) -> None:
        with self._lock:
            self._active = False

        if self._verbose:
            logging.info("QuickEntrySubscriber: deactivated")

    def is_active(self) -> bool:
        with self._lock:
            return self._active

    def get_accumulated_text(self) -> str:
        with self._lock:
            return self._finalized_text

    def get_display_text(self) -> str:
        with self._lock:
            if self._partial_text:
                return f"{self._finalized_text} {self._partial_text}".strip()
            return self._finalized_text

    def on_partial_update(self, result: 'RecognitionResult') -> None:
        with self._lock:
            if not self._active:
                return
            self._partial_text = result.text
            finalized = self._finalized_text
            partial = self._partial_text

        # Callback outside lock to prevent deadlocks
        self._on_text_change(finalized, partial)

        if self._verbose:
            logging.info(f"QuickEntrySubscriber: partial '{result.text}'")

    def on_finalization(self, result: 'RecognitionResult') -> None:
        with self._lock:
            if not self._active:
                return

            if self._finalized_text:
                self._finalized_text += " " + result.text
            else:
                self._finalized_text = result.text

            self._partial_text = ""
            finalized = self._finalized_text

        # Callback outside lock to prevent deadlocks
        self._on_text_change(finalized, "")

        if self._verbose:
            logging.info(f"QuickEntrySubscriber: finalized '{result.text}', "
                        f"total='{finalized}'")

File: src/client/winui/SttClient.Core/QuickEntry/QuickEntrySubscriber.cs
using Microsoft.Extensions.Logging;
using SttClient.Recognition;

namespace SttClient.QuickEntry;

/// <summary>
/// <see cref="IRecognitionSubscriber"/> that accumulates partial and finalized recognition
/// text while the QuickEntry popup is active.
///
/// Responsibilities:
/// - Gates accumulation behind an Activate/Deactivate lifecycle.
/// - Fires a display-text callback (outside the lock) whenever the visible text changes.
/// - Thread-safe: OnPartialUpdate and OnFinalization may be called from the ReceiveLoop task.
/// </summary>
public sealed class QuickEntrySubscriber : IRecognitionSubscriber
{
    private Action<string> _onTextChange;
    private readonly ILogger<QuickEntrySubscriber> _logger;
    private readonly object _lock = new();

    private string _finalizedText = string.Empty;
    private string _partialText = string.Empty;
    private bool _active;

    /// <summary>
    /// Initializes a new <see cref="QuickEntrySubscriber"/>.
    /// </summary>
    /// <param name="onTextChange">
    /// Callback invoked with the current display text (finalized + partial) on every update.
    /// Called outside the internal lock.
    /// </param>
    /// <param name="logger">Logger for diagnostic output.</param>
    public QuickEntrySubscriber(Action<string> onTextChange, ILogger<QuickEntrySubscriber> logger)
    {
        _onTextChange = onTextChange;
        _logger = logger;
    }

    /// <summary>
    /// Overrides the text-change callback. Used by <see cref="QuickEntryController"/> to wire
    /// the popup display update after the subscriber has been constructed.
    /// </summary>
    /// <param name="callback">New callback; replaces the one set at construction.</param>
    public void SetTextChangeCallback(Action<string> callback)
    {
        _onTextChange = callback;
    }

    /// <summary>Activates accumulation and clears any previously accumulated text.</summary>
    public void Activate()
    {
        lock (_lock)
        {
            _active = true;
            _finalizedText = string.Empty;
            _partialText = string.Empty;
        }

        _logger.LogInformation("QuickEntrySubscriber: activated");
    }

    /// <summary>Deactivates accumulation. Incoming results are ignored until next Activate.</summary>
    public void Deactivate()
    {
        lock (_lock)
        {
            _active = false;
        }

        _logger.LogInformation("QuickEntrySubscriber: deactivated");
    }

    /// <summary>Returns the accumulated finalized text (without the current partial).</summary>
    public string GetAccumulatedText()
    {
        lock (_lock)
        {
            return _finalizedText;
        }
    }

    /// <inheritdoc/>
    public void OnPartialUpdate(RecognitionResult result)
    {
        string displayText;

        lock (_lock)
        {
            if (!_active)
                return;

            _partialText = result.Text;
            displayText = BuildDisplayText(_finalizedText, _partialText);
        }

        _onTextChange(displayText);
        _logger.LogDebug("QuickEntrySubscriber: partial '{Text}'", result.Text);
    }

    /// <inheritdoc/>
    public void OnFinalization(RecognitionResult result)
    {
        string displayText;

        lock (_lock)
        {
            if (!_active)
                return;

            _finalizedText = _finalizedText.Length > 0
                ? _finalizedText + " " + result.Text
                : result.Text;
            _partialText = string.Empty;
            displayText = _finalizedText;
        }

        _onTextChange(displayText);
        _logger.LogDebug("QuickEntrySubscriber: finalized '{Text}', total='{Total}'", result.Text, displayText);
    }

    private static string BuildDisplayText(string finalized, string partial) =>
        partial.Length > 0
            ? (finalized.Length > 0 ? finalized + " " + partial : partial)
            : finalized;
}
